src/client.hpp

```diff
@@ -172,6 +182,38 @@ mtx::client::Client::post(
         do_request(session);
 }
 
+template<class Request>
+void
+mtx::client::Client::put(
+  const std::string &endpoint,
+  const Request &req,
+  std::function<void(std::experimental::optional<mtx::client::errors::ClientError>)> call,
```

Comment:
Better rename this to `callback` just to be consistent with the other methods.

Reply:
Renamed, and overloaded with a put that takes a request and response as input since some endpoints return data even on success. That doesn't follow the spec of PUT, however; I created an [issue in matrix-spec](https://github.com/matrix-org/matrix-doc/issues/1136), so if the API spec for those endpoints changes, we may be able to consolidate the put function into one. For now, both are necessary so the client can interface with synapse.